fabric-api-base/src/main/java/net/fabricmc/fabric/api/event/AutoInvokingEvent.java

```diff
@@ -0,0 +1,35 @@
+package net.fabricmc.fabric.api.event;
+
+import java.lang.annotation.*;
+
+/**
+ * Indicates that this {@link Event} is auto-invoking:
+ * it calls the event callback implemented by a context parameter type automatically and without registration.
+ *
+ * <p>This means that this event can be listened to in two ways:
+ * <ul>
+ *     <li>If the consumer is the context parameter and it implements the callback, it will be automatically invoked, don't register manually.
+ *     <li>Otherwise, there is no invocation and the listener needs manual registration as usual.
+ * </ul>
+ *
+ * <p>Do note that there may be more than one context parameter.
+ *
+ * <p>A typical use case is feature augmentation, for example to expose raw clicks to slots.
+ * The event callback has a slot parameter - the context parameter - and the event itself is carrying this annotation.
+ * All the slot needs to receive slot clicks is to implement {@code SlotClickCallback} on itself.
+ * It shouldn't do any explicit event registration like {@code SLOT_CLICK_EVENT.register(this::onSlotClick)},
+ * otherwise it will see extraneous callback invocations.
+ *
+ * <p>In general, an auto-invoking event bridges the gap between the flexibility of an event with global reach,
+ * and the convenience of implementing an interface that gets detected automatically.
+ *
+ * <p>This is a documentation-only annotation, the event factory has to implement the functionality explicitly by checking the parameter type and invoking it.
+ * On top of adding this annotation, the event field or method should document which parameters are context parameters,
+ * and under which circumstances they are invoked.
+ */
+// TODO: explore enforcing that auto-invoked listeners don't register themselves.
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
```

Comment:
Since this is just a decorative annotation, I don't see any point giving it a runtime retention.

Reply:
We thought it could possibly be used at runtime to perform some checks, so in doubt we decided to give it runtime retention. Now that you mention it I'm not 100% sure... Does this introduce a lot of bloat? I don't think a few annotated events will be a problem.

Reply:
It won't. After all it's an **annotation**, and keeping extra annotations won't kill people. Looks good.